### PR TITLE
Simulating trait data under multivariate Brownian diffusion

### PR DESCRIPTION
**Pull request by gabehassler**
Adding the `ParamsMultiBM` type for simulating multivariate Brownian diffusion on a tree/network. Shifts are not currently implemented and some functions could be made more memory efficient, but otherwise it should work for any situation where `ParamsBM` does.

### CODE REVIEW

**Comment by cecileane** [comment]
Thank you Gabe! Very nice organization (and thanks @pbastide for building the ParamsProcess type to be expandable, and for the recursion backbone) :smiley:

A few requests here, before merging:
- Could you possibly add an example of using your function, in the docstring for the function `simulate` perhaps? ideally in a `jldoctest` chunk, so that it's tested when the documentation
  is built. Also, it would be nice to give more indexing examples, such as  `sim[:InternalNodes, :Sim]` or `sim[:InternalNodes, :Exp]` or `sim[:All]`.
- Could you please add a warning somewhere that shifts are not implemented for the multiBM? in the docstring perhaps? or error thrown somewhere in the code?
- The function `siminds(params::ParamsBM, w::Symbol)` does not actually use the value `params`. (It only uses its type through the function's signature). So you could manifest this by removing the name "params", that is, define your function with: `siminds(::ParamsBM, w::Symbol)`. Same for the two methods of `preorderFunctions(::ParamsBM)` and `preorderFunctions(::ParamsMultiBM)`, and for `initSimulateBM(nodes::Vector{Node}, ::Tuple{ParamsBM})` (these functions don't use the value of `params`)
- It might be best to work on a branch, though not at all necessary (just fyi). We will "squash-and-merge", so this will create a single (new) commit on the master branch containing all your changes, and will cause your master branch to be out-of-sync with the main master branch. If you create a branch to contain all your new commit, and rewind your master branch to where it diverged from Claudia's, you won't need to rewrite your master branch's history to continue to track the main master branch.

Thanks Gabe!

Paul @pbastide: please review if wish and if you have time!

File: src/traits.jl [hunk 1]
```diff
@@ -978,49 +1046,77 @@ Sigma2: 0.1
...
                   checkPreorder=true::Bool)
     if isa(params, ParamsBM)
         model = "BM"
+    elseif isa(params, ParamsMultiBM)
+        model = "MBD"
     else
         error("The 'simulate' function only works for a BM process (for now).")
     end
     !ismissing(params.shift) || (params.shift = ShiftNet(net))
+
```
Comment (pbastide) [inline src/traits.jl:1078]: Looks like the only place where `ParamsBM` need to be mutable is here ? Or are there other places where we need it ? I think I agree with your first intuition that `ParamProcess` could be immutable though. Maybe there would be a way to make it so ? (But that might just be cosmetic).

File: test/test_simulate_mbd.jl [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+## Test of distibution
+###############################################################################
+@testset "Simulate test distribution" begin
+
+## Generate some values
+Random.seed!(18480224); # fix the seed
+μ = randn(trait_dim)
+Σ = randn(trait_dim, trait_dim)
+Σ = Σ * Σ' # needs to be positive definite
+pars = ParamsMultiBM(μ, Σ); # params of a MBD
```
Comment (pbastide) [inline test/test_simulate_mbd.jl:47]: Maybe add a test with a random root ? (So as to settle the earlier typo in root simulation)

Comment (cecileane) [reply test/test_simulate_mbd.jl:47]: that would be great indeed.

File: src/traits.jl (hunk 1 above)
Comment (gabehassler) [reply src/traits.jl:1078]: That was the only place I saw that requires it to be mutable. I think we could just build the `ShiftNet(net)` in the constructor for `ParamsBM`, which means both types could be immutable.

File: test/test_simulate_mbd.jl (hunk 2 above)
Comment (gabehassler) [reply test/test_simulate_mbd.jl:47]: Thanks for the suggestion! I'd noticed in the code coverage that the random root wasn't getting called, so planning to add that.

File: src/traits.jl [hunk 3]
```diff
@@ -1033,6 +1129,23 @@ function updateRootSimulateBM!(M::Matrix, i::Int, params::Tuple{ParamsBM})
...
+                                i::Int,
+                                params::Tuple{ParamsMultiBM})
+    params = params[1]
+    p = process_dim(params)
+
+    means, vals = partitionMBDMatrix(M, p)
+
+    if (params.randomRoot)
+        means[:, i] .= params.mu # expectation
+        vals[:, i] .= params.mu + params.L * randn(p) # random value #TODO: make memory efficient
```
[the remark this answers is not in the document]
Comment (gabehassler) [reply src/traits.jl:1142]: Good point, thanks for catching that. Given that we only need it once, I think we can just compute it here and don't need to store it in `ParamsMultiBM`.

**Comment by gabehassler** [comment]
Thanks @cecileane and @pbastide for the thoughtful feedback. I'll make the changes then submit another pull request.

Regarding the `ShiftNet`, I don't think it would be too challenging to redefine the structure to store a `Matrix{Float64}` for shifts rather than a `Vector{Float64}`. As far as I can tell, the only thing that would need to change to make the existing code work with the new internal structure is to change the `ShiftNet` constructors.

@pbastide suggested we could simply make `ParamsBM` multivariate by default (and cautioned that this may create issues with memory efficiency). If others feel it's better to avoid some code duplication and simply have one class for a (potentially multivariate) Brownian diffusion process on a tree, I can rewrite the matrix operations to be memory efficient (which I should do anyway).

**Comment by cecileane** [comment]
About making `paramsBM` multivariate by default: I don't feel strongly one way or another. One one hand, specialized methods can be made more efficient. On the other hand, code duplication isn't great. In any case, I like your `preorderFunctions`, because it could make it easier to extend to more models later.

**Pull request closed without merging**